src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/CompositeModelBinder.cs

```diff
@@ -76,12 +76,13 @@ public virtual Task<ModelBindingResult> BindModelAsync(ModelBindingContext bindi
                             ValidationStateEntry entry;
                             if (!bindingContext.ValidationState.TryGetValue(result.Model, out entry))
                             {
-                                entry = new ValidationStateEntry();
+                                entry = new ValidationStateEntry()
+                                {
+                                    Key = result.Key,
+                                    Metadata = bindingContext.ModelMetadata,
+                                };
                                 bindingContext.ValidationState.Add(result.Model, entry);
                             }
-
-                            entry.Key = entry.Key ?? result.Key;
-                            entry.Metadata = entry.Metadata ?? bindingContext.ModelMetadata;
```

Comment:
I don't think this was good to have. This was fighting with placed in code where we want to have a `null` key on purpose.

The `ValidationVisitor` already knows how to compensate for incomplete entries, so this wasn't helping.